### DISCUSSION TITLE
Authenticating and Updating Cookie Based Session

### DISCUSSION THREAD

**Discussion by kevinswarner** [remix-run/remix#4517] [Q&A]
I am very new to Remix, so this may be obvious. I just cannot figure out the best solution.

I want to protect a several pages available only to authenticated users. So for example, lets say I have a page "profile.tsx". I need to perform an auth check when someone visits this page, and if not authenticated, redirect them to the sign in page. I want to authenticate using a JWT, which has been saved in the session using the cookie session storage. Also, if the response I get back from my authentication provider (Stytch) includes a new (refreshed) JWT, I want to update that value in my current session.

My approach was to use a loader on the page, and within the loader, make the call to Stytch. The library Stytch uses ONLY actually makes a remote fetch if the JWT has expired, otherwise, the user is considered authenticated. However, if the JWT is expired, it makes a remove fetch to authenticate and get a new JWT.

My problem is that I cannot figure out how to update the JWT in the current session / cookie. Here is my loader right now (simplified)...

```typescript
export async function loader({ request }: LoaderArgs) {
  const session = await getSession(request);
  const jwt = session.get('jwt');

  try {
    const authenticationResponse = await client.sessions.authenticateJwt(jwt);
    if (jwt !== authenticationRepsonse.session_jwt) {
      // This is where I want to update the JWT stored in the session cookie.
      // I tried this, but I don't think it is ever getting persisted to the cookie.
      session.set('jwt', authenticationResponse.session_jwt);
      await sessionStorage.commitSession(session);
      // Now, I just want to continue on to the route as normal.
    }
  } catch (error) {
    // Invalid session. This is where I want to redirect to the sign in page.
    return redirect('/signin');
  }
}
```

I see in examples how to set the cookie when redirecting somewhere, but I really don't want to redirect. I just want to have the cookie be updated. Also, I actually would move most of this code to a shared session.server.ts file, but I simplified my example here.

I appreciate any advice!

**Comment by machour** [comment]
You almost figured it out!

In order to get persisted, the cookie needs to be sent back to the browser.

```js
await sessionStorage.commitSession(session);
```

returns the value that needs to be set on the `Set-Cookie` header of your loader's `Response`.
So instead of ignoring its result like you're doing now, you'll have to use it like this:

```js
  return json(
    { yourData },
    {
      headers: {
        "Set-Cookie": await sessionStorage.commitSession(),
      },
    }
  );
```

That should do the trick

**Comment by kevinswarner** [comment]
Thanks @machour! But, now, if my code actually lives in a separate file and is called from my loader, I am a little confused about what I would do.

```typescript
// session.server.ts

export async function authenticateSessios(request) {
  const session = await getSession(request);
  const jwt = session.get('session_jwt');

  try {
    const authenticationResponse = await client.sessions.authenticateJwt(jwt);
    if (jwt !== authenticationResponse.session_jwt) {
      session.set('session_jwt', authenticationResponse.session_jwt);
      await sessionStorage.commitSession(session);

      // Is it possible to set cookie header here somehow?
      
    }
    return session;
  } catch (error) {
    console.log('Invalid session.')
    throw redirect('/signin');
  }
}
```

And my loader...

```typescript
// profile.tsx
export async function loader({ request }: LoaderArgs) {
  const session = await authenticateSession(request);
  
  // Do other stuff here...

}
```

I want to have this authentication code in a separate file/function because it will need to be called from every protected route that I have in my app, which will be quite a few.

Is there a way I can "save" to the cookie from within the session.server.ts file or must I do it in the loader function as you pointed out?

Thanks!

**Reply by kevinswarner** [reply]
I followed the example in this discussion. It seems to work. https://github.com/remix-run/remix/discussions/2166
